### PR TITLE
ChangeLog Feature

### PR DESCRIPTION
**Pull request by Draven714**
Hello,

I am creating this pull request on the off chance you're interested in including this feature.

If the USE_CHANGELOG variable is set to true then the gitlog-to-changelog.sh file will generate a CHANGELOG file using the git log messages from the last version tag up to HEAD.

I'm sure it's not perfect, but I've never written a shell script before.
I can also see where it might not work if the user deletes all their tags.

And the ChangeLog code should be moved from the format-tag-message.sh file into it's own file but I didn't want to go too far with your code.

### CODE REVIEW

**Comment by jaspernbrouwer** [comment]
Thanks for submitting a PR! The feature sounds interesting, I'll look into it shortly! (I hope)

**Comment by Draven714** [comment]
Great! I look forward to hearing back about it :)

**Comment by dhilgarth** [comment]
I think this feature needs a bit of refinement. The [sample](https://github.com/Draven714/git-flow-hooks/blob/master/CHANGELOG) you provide already shows that these change logs will contain stuff that nobody would really put into their changelog file.  
Examples are:
+ * Bumped version to 1.0.7 (Michael Delle)
 Messages about merged branches. Wouldn't it be more helpful to display the actual commit messages from the merged branches?

**Comment by Draven714** [comment]
I very much agree.

File: modules/format-tag-message.sh [hunk 1]
```diff
@@ -1,5 +1,13 @@
 #!/usr/bin/env bash
 
+# Move to own file.
+USE_CHANGELOG=$(__get_use_changelog)
+if [ "$USE_CHANGELOG" = true ]; then
+    VERSION_CURRENT=$(__get_current_version)
+    CHANGELOG_FILE_NAME=$(__get_changelog_file_name)
+    CHANGE=$($HOOKS_DIR/modules/gitlog-to-changelog.sh $VERSION_CURRENT $VERSION "$CHANGELOG_FILE_NAME")
+fi
+
```
Comment (jaspernbrouwer) [inline modules/format-tag-message.sh]: I agree with the comment "Move to own file" here :)

File: modules/git-flow-hooks-config.sh.dist [hunk 2]
```diff
@@ -3,3 +3,5 @@ VERSION_BUMPLEVEL_HOTFIX="PATCH"
 VERSION_BUMPLEVEL_RELEASE="MINOR"
 VERSION_SORT="/usr/bin/sort -V"
 VERSION_TAG_PLACEHOLDER="%tag%"
+USE_CHANGELOG=false
+CHANGELOG_FILE_NAME="CHANGELOG"
```
Comment (jaspernbrouwer) [inline modules/git-flow-hooks-config.sh.dist]: I'd prefer a common prefix to these variables, in this case `CHANGELOG_`.
Also, `CHANGELOG_FILE` would be more in line with `VERSION_FILE`.

Could you change the settings to:

```sh
CHANGELOG_ENABLE=false
CHANGELOG_FILE="CHANGELOG"
```

File: README.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ What does it do?
 - Prevent merge marker commits.
 - Automatically bump versions when starting a release or hotfix. Versions are generated, written to file and committed.
 - Automatically specify tag messages.
+- Automatically creates a CHANGELOG file if USE_CHANGELOG=true.
```
Comment (jaspernbrouwer) [inline README.md]: Maybe be a bit less specific about the technical side and simple state what it does?

I think "Automatically create a change-log." would suffice.

File: modules/git-flow-hooks-config.sh.dist (hunk 2 above)
Comment (Draven714) [inline modules/git-flow-hooks-config.sh.dist]: I was thinking you'd take over the feature if you wanted to include it.
Writing shell scripts isn't what I do, I just wanted this feature for my projects.

But if you want me to refine it first, I can do that.

**Comment by jaspernbrouwer** [comment]
@Draven714: As you can see I had some comments on your changes.

Besides those, can you also remove the `CHANGELOG` and `VERSION` files from this PR?

And when you're done, maybe rebase and squash the whole PR? All these commits would make the merge a bit messy.

I also agree that certain commits can be omitted from the change-log, like the version bumps and merge commits. There's a `VERSION_BUMP_MESSAGE` setting now, which you should use to filter out the version bumps. Merge commits can be a bit more tricky, but maybe you can do something with the `gitflow.hotfix.finish.message` en `gitflow.release.finish.message` options?

PS: Although I didn't get around to diving into this PR earlier, I still think this would be an awesome feature! You're work is very much appreciated!

**Comment by Draven714** [comment]
Whoops, commented on one of your earlier comments about 30 seconds too soon.

To reiterate, I don't normally write shell scripts, my first time changing more then one line I think :)

Most of the changes you want done I can probably figure out.

**Comment by jaspernbrouwer** [comment]
I'd appreciate it if you would give it a try! To be honest, shell-scripting isn't my thing either :)

If you run into problems, please state them, and I'll take over.

**Comment by accessd** [comment]
@Draven714 Do you still working on this feature?

**Comment by jaspernbrouwer** [comment]
I finally ~~have~~ took some time to scrub issues and PRs and see this is still open. There hasn't been any progress for half a year now, no hard feelings of course!

I'm closing this PR because I too don't have the time to test and refine this feature. I'm open to it, but really need help with the actual work.

**Pull request closed without merging**